File: StructuralPatterns/Facade/facade.py
# Subsystem components
class CPU:
    def start(self) -> str:
        """
        Start the CPU.

        Returns:
            str: A message indicating the CPU has started.
        """
        return "CPU started"


class Memory:
    def load(self) -> str:
        """
        Load memory.

        Returns:
            str: A message indicating that memory has been loaded.
        """
        return "Memory loaded"


class HardDrive:
    def read(self) -> str:
        """
        Read from the hard drive.

        Returns:
            str: A message indicating the hard drive has been read.
        """
        return "Hard Drive read"


# Facade
class ComputerFacade:
    def __init__(self):
        """
        Initialize the ComputerFacade with CPU, Memory, and HardDrive instances.
        """
        self.cpu: CPU = CPU()
        self.memory: Memory = Memory()
        self.hard_drive: HardDrive = HardDrive()

    def start(self) -> list[str]:
        """
        Start the computer by orchestrating the startup sequence.

        Returns:
            list[str]: A list of messages indicating each step of the startup sequence.
        """
        result: list[str] = [self.cpu.start(), self.memory.load(), self.hard_drive.read()]
        return result


# Client code
if __name__ == "__main__":
    # Using the Facade to start the computer
    computer_facade: ComputerFacade = ComputerFacade()
    result: list[str] = computer_facade.start()

    # Displaying the result
    print("Computer Starting Sequence:")
    for step in result:
        print(step)

File: StructuralPatterns/Facade/facade.cpp
#include <iostream>
#include <vector>

// Subsystem components
class CPU {
public:
    std::string start() {
        return "CPU started";
    }
};

class Memory {
public:
    std::string load() {
        return "Memory loaded";
    }
};

class HardDrive {
public:
    std::string read() {
        return "Hard Drive read";
    }
};

// Facade
class ComputerFacade {
private:
    CPU cpu;
    Memory memory;
    HardDrive hardDrive;

public:
    std::vector<std::string> start() {
        std::vector<std::string> result;
        result.push_back(cpu.start());
        result.push_back(memory.load());
        result.push_back(hardDrive.read());
        return result;
    }
};

// Client code
int main() {
    // Using the Facade to start the computer
    ComputerFacade computerFacade;

    // Starting the computer
    std::vector<std::string> result = computerFacade.start();

    // Displaying the result
    std::cout << "Computer Starting Sequence:" << std::endl;
    for (const auto& step : result) {
        std::cout << step << std::endl;
    }

    return 0;
}
